Generate Server Code

amberchow8 opened the pull request:
-Added short-unique-id library 
-Added middleware to Study schema where server code is generated when a Study is created in the database
-Changed server code to serverCode property of the Study on the StudyBuilderPage
-Closes #82 

*I couldn't figure out how to update current studies in the database with the new serverCode property but I created new studies on the UI and their server codes show up

amberchow8 commented on packages/api/src/models/study.ts:
```diff
@@ -36,6 +40,13 @@ const studySchema = new Schema<IStudy>({
 
+studySchema.pre("save", async function (next) {
+  if (!this.serverCode) {
+    this.serverCode = uid.rnd(5);
```
Not sure if length of 5 is long enough

ross3102 commented on packages/ui/src/stores/studyBuilder.ts:
```diff
@@ -192,6 +195,7 @@ export const useStudyBuilderStore = defineStore("studyBuilder", () => {
       description: description.value ?? "",
       batteries: taskBank.value.map((id) => taskData.value[id]), // TODO: fix this
       sessions: sessions.value.map((id) => sessionData.value[id]),
+      serverCode: serverCode.value!,
```
can this be undefined? curious why the `!`

ross3102 reviewed (approved):
looks good! just small comments

amberchow8 commented:
I changed the pre middleware function a little to regenerate the server code if the first one generated is found in a study. I also added the alphanumeric lowercase dictionary for code generation so all letters are lowercase.

ross3102 commented on packages/api/src/models/study.ts:
```diff
@@ -36,6 +41,18 @@ const studySchema = new Schema<IStudy>({
+
+    const study = await Study.findOne({ serverCode: this.serverCode });
+    if (study) {
+      this.serverCode = uid.rnd(serverCodeLength);
```
good approach, but this will need to be a while loop of some kind in case the regenerated code is also taken

ross3102 reviewed (changes requested):
all good except making sure the code generation never overlaps!

ross3102 reviewed (approved):
LGTM!
